### PR TITLE
Removes WPPostContentViewProvider from PostListViewController

### PR DESCRIPTION
**Pull request by diegoreymendez**
First step towards resolving #5250 

### Details:

Removes `WPPostContentViewProvider` from `PostListViewController`, and simplifies the code a bit.

I've added a few temporary elements which I plan to remove in the future.  The purpose of these additions was to make the code a bit more readable - at least until I can figure out a better solution.

These new elements are:

- `ConfigurablePostView` protocol.
- `InteractivePostView` protocol.
- `InteractivePostViewDelegate` protocol.

### Testing:

1. Run the app.
2. Go to the list of posts for a blog.
3. Try these features: create post, edit post, view post, view stats, trash post, restore post, delete post.

Needs review: @jleandroperez

### CODE REVIEW

**Comment by aerych** [comment]
@diegoreymendez I'm a fan of improving WPContentProvider or replacing it with something better.  One thing it was doing for us was decoupling our model and view layer.  It looks like these changes will more tightly couple cells and the post model and that seems counter productive. Could things be tweaked to keep the view and model separated?

**Comment by diegoreymendez** [comment]
@aerych - The main reasons why I'm pushing these changes:

1. This protocol was not being successful at decoupling and was actually complicating the code more than it was being useful.  [Here's a very clear example](https://github.com/wordpress-mobile/WordPress-iOS/pull/5391/files#diff-56d3ee28947d056476717395cb71f4d7R593) of how it's both not really decoupling and overcomplicating the code simultaneously. (ie: having to cast a provider back to a post is a code smell)

2. `WPPostContentViewProvider` can be roughly translated as "post content provider".  It was basically an alias for regular post objects.  Why would post objects need a provider for their own data, which they're still storing?  This would make sense if the protocol was representing different objects, but it isn't.

3. Taking into account points 1 and 2.  The kind of decoupling we could take advantage of is one with different storage than CoreData objects (as in `Object` vs `RemoteObject` from services).  Even then I'm not sure if we'd gain a lot by going that route.  However, this PR is intentionally only meant to simplify and cleanup what we have right now.

Last but not least, there's the following question regarding the original idea of decoupling: for what purpose are we trying to decouple VC from `AbstractPost`?

This is not a rhetorical question - the goal's definitely not clear to me.

**Comment by diegoreymendez** [comment]
@aerych - By the way... if we want a decoupling solution as mentioned in point 3 I'll be glad to giving it a shot.  That makes a bit more sense in my mind because it's trully decoupling the data that the views consume, from the model data.

Let me know if that sounds good and I'll start tackling it in a separate discovery PR.

**Comment by aerych** [comment]
I believe the original idea, the one that started the whole content provider thing, came from Mike as a strategy to decouple a view from the model it represents, and to let one type of view display different kinds of content.  From an MVC standpoint, this is a good thing and (in theory) would let us unit test a view independent of the model. In practice we never quite got there.  I think our views might be a bit too complicated for that.

While the protocol was at least hiding the model from the view, the changes here give up any pretense at decoupling. This isn't really fixing things, its trading a poor implementation for a bad practice.  I completely agree that the content provider protocol has gotten a bit messy, and I'm totally in favor of a clean up, but a view storing a reference to its model is a code smell of its own.

I'd suggest a slight tweak to the approach here.  Rather than storing the model in the view, instead store a some other identifier, perhaps the NSManagedObjectID, that the controller could use to identify and retrieve the corresponding model object. `InteractivePostViewDelegate` could then pass that identifier.  This would help to keep the model and the view decoupled.   I realize this might feel like complicating things but its proper for the view to be ignorant of the model. 

>  The kind of decoupling we could take advantage of is one with different storage than CoreData objects (as in Object vs RemoteObject from services). Even then I'm not sure if we'd gain a lot by going that route.

I'm not sure that's the kind of decoupling we need, and I agree that we wouldn't gain much going that route.  
In any event, I do like the strict separate between the view and model so passing a model directly to a view for configuration purposes makes me cringe a bit (strictly speaking controllers should be configuring the views) and we loose our theoretical ability to unit test the view independent of the model.  That said, it does simplify things and is probably fine until we find a better solution. 

Thoughts?

**Comment by aerych** [comment]
I missed your other comment while I was drafting my own :)

>  That makes a bit more sense in my mind because it's trully decoupling the data that the views consume, from the model data

I think I'm still suffering PTSD from the over-architected view-model in the old `LoginViewController` so anything that sounds like a view-model makes me shudder. ;)  But it would still be interesting to explore.

**Comment by diegoreymendez** [comment]
> I think I'm still suffering PTSD from the over-architected view-model in the old LoginViewController so anything that sounds like a view-model makes me shudder. ;) But it would still be interesting to explore.

I'll draft an example up in the next few days and ping you.

File: WordPress/Classes/ViewRelated/Post/InteractivePostViewDelegate.h [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+///
+@protocol InteractivePostViewDelegate <NSObject>
+@optional
+- (void)cell:(nonnull UITableViewCell *)cell handleEditPost:(nonnull AbstractPost *)post;
+- (void)cell:(nonnull UITableViewCell *)cell handleViewPost:(nonnull AbstractPost *)post;
+- (void)cell:(nonnull UITableViewCell *)cell handleStatsForPost:(nonnull AbstractPost *)post;
+- (void)cell:(nonnull UITableViewCell *)cell handleTrashPost:(nonnull AbstractPost *)post;
+- (void)cell:(nonnull UITableViewCell *)cell handlePublishPost:(nonnull AbstractPost *)post;
+- (void)cell:(nonnull UITableViewCell *)cell handleRestorePost:(nonnull AbstractPost *)post;
+@end
```
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Post/InteractivePostViewDelegate.h]: Nipticky: Missing newline

File: WordPress/Classes/ViewRelated/Post/PostListViewController.swift [hunk 2]
```diff
@@ -586,72 +591,30 @@ class PostListViewController : AbstractPostListViewController, UIViewControllerR
...
-        }
+    // MARK: - InteractivePostViewDelegate
 
-        editPost(apost)
+    func cell(cell: UITableViewCell, handleEditPost post: AbstractPost) {
+        editPost(post)
     }
 
-    func cell(cell: UITableViewCell!, receivedViewActionForProvider contentProvider: WPPostContentViewProvider!) {
-        guard let apost = contentProvider as? AbstractPost else {
```
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Post/PostListViewController.swift]: Glad to see this snippet go!

**Comment by jleandroperez** [comment]
@diegoreymendez all tests pass on the device.

:shipit: on my side, code looks good.

File: WordPress/Classes/ViewRelated/Post/InteractivePostViewDelegate.h (hunk 1 above)
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Post/InteractivePostViewDelegate.h]: Added.

**Comment by diegoreymendez** [comment]
Thanks Jorge!

**Pull request merged**